tests/Microsoft.Identity.Web.Test/WebAppExtensionsTests.cs

```diff
@@ -1,32 +1,493 @@
+        {
+            var httpContext = HttpContextUtilities.CreateHttpContext();
+            httpContext.RequestServices = provider;
+            var authScheme = new AuthenticationScheme(OpenIdConnectDefaults.AuthenticationScheme, OpenIdConnectDefaults.AuthenticationScheme, typeof(OpenIdConnectHandler));
+            var authProperties = new AuthenticationProperties();
+            var authCodeReceivedContext = new AuthorizationCodeReceivedContext(httpContext, authScheme, oidcOptions, authProperties);
+
+            await oidcOptions.Events.AuthorizationCodeReceived(authCodeReceivedContext).ConfigureAwait(false);
+
+            // Assert event called
+            await authCodeReceivedFuncMock.ReceivedWithAnyArgs().Invoke(Arg.Any<AuthorizationCodeReceivedContext>()).ConfigureAwait(false);
+            await ((ITokenAcquisitionInternal)tokenAcquisitionMock).ReceivedWithAnyArgs().AddAccountToCacheFromAuthorizationCodeAsync(Arg.Any<AuthorizationCodeReceivedContext>(), Arg.Any<IEnumerable<string>>()).ConfigureAwait(false);
+        }
+
+        private async Task AddWebAppCallsProtectedWebApi_TestTokenValidatedEvent(OpenIdConnectOptions oidcOptions, Func<TokenValidatedContext, Task> tokenValidatedFuncMock)
+        {
+            var httpContext = HttpContextUtilities.CreateHttpContext();
+            httpContext.Request.Form = new FormCollection(
+                new Dictionary<string, StringValues>() { { ClaimConstants.ClientInfo, Base64UrlHelpers.Encode($"{{\"uid\":\"{TestConstants.TenantIdAsGuid}\",\"utid\":\"{TestConstants.TenantIdAsGuid}\"}}") } });
+            var authScheme = new AuthenticationScheme(OpenIdConnectDefaults.AuthenticationScheme, OpenIdConnectDefaults.AuthenticationScheme, typeof(OpenIdConnectHandler));
+            var authProperties = new AuthenticationProperties();
+            var claimsPrincipal = new ClaimsPrincipal();
+            var tokenValidatedContext = new TokenValidatedContext(httpContext, authScheme, oidcOptions, httpContext.User, authProperties);
+
+            await oidcOptions.Events.TokenValidated(tokenValidatedContext).ConfigureAwait(false);
+
+            // Assert event called, properties set
+            await tokenValidatedFuncMock.ReceivedWithAnyArgs().Invoke(Arg.Any<TokenValidatedContext>()).ConfigureAwait(false);
+            Assert.True(tokenValidatedContext.Principal.HasClaim(c => c.Type == ClaimConstants.Tid));
+            Assert.True(tokenValidatedContext.Principal.HasClaim(c => c.Type == ClaimConstants.UniqueObjectIdentifier));
+        }
+
+        private async Task AddWebAppCallsProtectedWebApi_TestRedirectToIdentityProviderForSignOutEvent(IServiceProvider provider, OpenIdConnectOptions oidcOptions, Func<RedirectContext, Task> redirectFuncMock, ITokenAcquisition tokenAcquisitionMock)
```

Comment:
nit: These last three have similar code set up, could you move some of it to another private method to reduce the code duplication? then there is less noise in the differences between the methods.

Reply:
I second @jennyf19 that we should avoid duplication.
@pmaytak there is a took in Visual Studio named "Code Clone" which will tell you which code is copied/paste. you might want to use it.